### PR TITLE
Added regex support to the `split` expression

### PR DESCRIPTION
**Pull request by DeltaRays**
---
**Target Minecraft Versions:** any    
**Requirements:**   none  
**Related Issues:**   none

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java [hunk 1]
```diff
@@ -1,98 +1,102 @@
-/**
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java:1]: Line endings for history / viewing the changes properly 😃

File: src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java [hunk 2]
```diff
@@ -1,98 +1,102 @@
...
+	}
+	
+	@Override
+	public Class<? extends String> getReturnType() {
+		return String.class;
+	}
+	
+	@Override
+	public String toString(final @Nullable Event e, final boolean debug) {
+		return join ? "join " + strings.toString(e, debug) + (delimiter != null ? " with " + delimiter.toString(e, debug) : "") : ((regex ? "regex " : "")+"split " + strings.toString(e, debug) + (delimiter != null ? " at " + delimiter.toString(e, debug) : ""));
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java:99]: To avoid having a long line, could you split this up over a few lines?

File: src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java [hunk 3]
```diff
@@ -1,98 +1,102 @@
...
+/**
+ * @author Peter Güttinger
+ */
+@Name("Join & Split")
+@Description("Joins several texts with a common delimiter (e.g. \", \"), or splits a text into multiple texts at a given delimiter.")
+@Examples({"message \"Online players: %join all players with \"\" | \"\"%\" # %all players% would use the default \"x, y, and z\"",
+	"set {_s::*} to the string argument split at \",\""})
+@Since("2.1, INSERT VERSION (added regex support)")
+public class ExprJoinSplit extends SimpleExpression<String> {
+	static {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java:48]: Spaces before and after static declaration

File: src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java (hunk 1 above)
Comment (DeltaRays) [reply src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java:1]: I don't know why but this is always an issue

**Comment by DeltaRays** [comment]
Hopefully fixed the new line issues

**Comment by ThatOneTqnk** [comment]
Might want to try `dos2unix` on the source file, seems the line endings were changed from CRLF to LF

**Comment by ThatOneTqnk** [comment]
The issue may be your `core.autocrlf` setting for Git, a few options will make sure the line endings are changed to LF when writing the file to the object database.

**Comment by DeltaRays** [comment]
Okay, I changed it to false

File: src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java (hunk 1 above)
Comment (Matocolotoe) [reply src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java:1]: Because line endings are not consistent in Skript's files : some are LF and some are CRLF, it's quite random

**Review by Wealthyturtle — DISMISSED** [review]
Line endings are still not fixed

**Comment by DeltaRays** [comment]
It returns an error when I change line endings

**Comment by DeltaRays** [comment]
It really doesn't want to change

**Comment by DeltaRays** [comment]
I really don't know why it doesn't work

**Comment by DeltaRays** [comment]
Okay, my last pull request should fix all those issues

**Review by APickledWalrus — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java [hunk 4]
```diff
@@ -42,16 +42,23 @@
 @Name("Join & Split")
 @Description("Joins several texts with a common delimiter (e.g. \", \"), or splits a text into multiple texts at a given delimiter.")
 @Examples({"message \"Online players: %join all players with \"\" | \"\"%\" # %all players% would use the default \"x, y, and z\"",
-		"set {_s::*} to the string argument split at \",\""})
-@Since("2.1")
+	"set {_s::*} to the string argument split at \",\""})
+@Since("2.1, INSERT VERSION (added regex support)")
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java:46]: "added" isn't really needed here.
```suggestion
@Since("2.1, INSERT VERSION (regex support)")

```

File: src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java [hunk 5]
```diff
@@ -92,7 +100,13 @@ public boolean isSingle() {
...
 	@Override
 	public String toString(final @Nullable Event e, final boolean debug) {
-		return join ? "join " + strings.toString(e, debug) + (delimiter != null ? " with " + delimiter.toString(e, debug) : "") : "split " + strings.toString(e, debug) + (delimiter != null ? " at " + delimiter.toString(e, debug) : "");
+		return join ? "join " + strings.toString(e, debug)
+			+(delimiter != null ? " with "
+			+ delimiter.toString(e, debug) : "")
+			: ((regex ? "regex " : "")
+			+"split " + strings.toString(e, debug)
+			+ (delimiter != null ? " at "
+			+ delimiter.toString(e, debug) : ""));
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java:109]: This was an unnecessary line change.
```suggestion
		return join ? "join " + strings.toString(e, debug) +(delimiter != null ? " with " + delimiter.toString(e, debug) : "") : ((regex ? "regex " : "") +"split " + strings.toString(e, debug) + (delimiter != null ? " at " + delimiter.toString(e, debug) : ""));

```

File: src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java [hunk 6]
```diff
@@ -42,16 +42,23 @@
...
+			"(concat[enate]|join) %strings% [(with|using|by) [[the] delimiter] %-string%]",
+			"split %string% (at|using|by) [[the] delimiter] %string%",
+			"%string% split (at|using|by) [[the] delimiter] %string%",
+			"regex split %string% (at|using|by) [[the] delimiter] %string%",
+			"regex %string% split (at|using|by) [[the] delimiter] %string%");
 	}
 	
+	private boolean regex;
+	
 	private boolean join;
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java:60]: 1) The extra line is unneeded
2) Since regex is assigned after join, it should be below join here
```suggestion
	private boolean join;
	private boolean regex;

```

File: src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java [hunk 7]
```diff
@@ -42,16 +42,23 @@
 @Name("Join & Split")
 @Description("Joins several texts with a common delimiter (e.g. \", \"), or splits a text into multiple texts at a given delimiter.")
 @Examples({"message \"Online players: %join all players with \"\" | \"\"%\" # %all players% would use the default \"x, y, and z\"",
-		"set {_s::*} to the string argument split at \",\""})
-@Since("2.1")
+	"set {_s::*} to the string argument split at \",\""})
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java:45]: This was an unnecessary change
```suggestion
		"set {_s::*} to the string argument split at \",\""})

```

**Review by ShaneBeee — CHANGES_REQUESTED** [review]
Just a few minor requests, mainly to bring this class back to how it was before the unnecessary changes.

File: src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java [hunk 8]
```diff
@@ -42,15 +42,24 @@
...
+			"split %string% (at|using|by) [[the] delimiter] %string%",
+			"%string% split (at|using|by) [[the] delimiter] %string%",
+			"regex split %string% (at|using|by) [[the] delimiter] %string%",
+			"regex %string% split (at|using|by) [[the] delimiter] %string%");
 	}
 	
 	private boolean join;
+	private boolean regex;
+
+
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java:62]: no need for the extra line here.

File: src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java [hunk 9]
```diff
@@ -91,7 +101,8 @@ public boolean isSingle() {
 	
 	@Override
 	public String toString(final @Nullable Event e, final boolean debug) {
-		return join ? "join " + strings.toString(e, debug) + (delimiter != null ? " with " + delimiter.toString(e, debug) : "") : "split " + strings.toString(e, debug) + (delimiter != null ? " at " + delimiter.toString(e, debug) : "");
+		return join ? "join " + strings.toString(e, debug) +(delimiter != null ? " with " + delimiter.toString(e, debug) : "") : ((regex ? "regex " : "") +"split " + strings.toString(e, debug) + (delimiter != null ? " at " + delimiter.toString(e, debug) : ""));
+
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java:105]: no need for the extra line here

File: src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java [hunk 10]
```diff
@@ -91,7 +101,8 @@ public boolean isSingle() {
 	
 	@Override
 	public String toString(final @Nullable Event e, final boolean debug) {
-		return join ? "join " + strings.toString(e, debug) + (delimiter != null ? " with " + delimiter.toString(e, debug) : "") : "split " + strings.toString(e, debug) + (delimiter != null ? " at " + delimiter.toString(e, debug) : "");
+		return join ? "join " + strings.toString(e, debug) +(delimiter != null ? " with " + delimiter.toString(e, debug) : "") : ((regex ? "regex " : "") +"split " + strings.toString(e, debug) + (delimiter != null ? " at " + delimiter.toString(e, debug) : ""));
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java:104]: Just some spacing issues.
```suggestion
		return join ? "join " + strings.toString(e, debug) + (delimiter != null ? " with " + delimiter.toString(e, debug) : "") : ((regex ? "regex " : "") + "split " + strings.toString(e, debug) + (delimiter != null ? " at " + delimiter.toString(e, debug) : ""));
```

**Review by ShaneBeee — CHANGES_REQUESTED** [review]
Just some small changes

File: src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java [hunk 11]
```diff
@@ -42,15 +42,23 @@
 @Description("Joins several texts with a common delimiter (e.g. \", \"), or splits a text into multiple texts at a given delimiter.")
 @Examples({"message \"Online players: %join all players with \"\" | \"\"%\" # %all players% would use the default \"x, y, and z\"",
 		"set {_s::*} to the string argument split at \",\""})
-@Since("2.1")
+
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprJoinSplit.java:45]: Im not sure if I didn't notice this before, or its new, but there shouldn't be a space here

**Review by ShaneBeee — CHANGES_REQUESTED** [review]
Just a small one I missed before.

**Review by ShaneBeee — APPROVED** [review]
